src/transformers/models/pop2piano/feature_extraction_pop2piano.py

```diff
@@ -0,0 +1,390 @@
+                )
+            warnings.warn("return_attention_mask is set to True for batched inputs")
+            return_attention_mask = True
+        else:
+            # To process it in the same pipeline
+            audio = [audio]
+            sampling_rate = [sampling_rate]
+
+        batch_input_feature, batch_beatsteps, batch_ext_beatstep = [], [], []
+        for single_raw_audio, single_sampling_rate in zip(audio, sampling_rate):
+            # If it's [np.ndarray]
+            if isinstance(single_raw_audio, list) and isinstance(single_raw_audio[0], np.ndarray):
+                single_raw_audio = single_raw_audio[0]
+
+            bpm, beat_times, confidence, estimates, essentia_beat_intervals = self.extract_rhythm(
+                audio=single_raw_audio
+            )
+            beatsteps = self.interpolate_beat_times(beat_times, steps_per_beat, extend=True)
+
+            if self.sampling_rate != single_sampling_rate and self.sampling_rate is not None:
+                # Change sampling_rate to self.sampling_rate
+                single_raw_audio = librosa.core.resample(
+                    single_raw_audio,
+                    orig_sr=single_sampling_rate,
+                    target_sr=self.sampling_rate,
+                    res_type="kaiser_best",
+                )
+
+            single_sampling_rate = self.sampling_rate
+            start_sample = int(beatsteps[0] * single_sampling_rate)
+            end_sample = int(beatsteps[-1] * single_sampling_rate)
+
+            input_feature, extrapolated_beatstep = self.single_preprocess(
+                beatstep=beatsteps - beatsteps[0],
+                audio=torch.from_numpy(single_raw_audio)[start_sample:end_sample],
+            )
+
+            input_feature = np.transpose(self.log_mel_spectogram(input_feature), (0, -1, -2))
+
+            batch_input_feature.append(input_feature)
+            batch_beatsteps.append(beatsteps)
+            batch_ext_beatstep.append(extrapolated_beatstep)
+
+        if return_attention_mask:
+            output = BatchFeature(
+                {
+                    "input_features": batch_input_feature,
+                    "beatsteps": batch_beatsteps,
+                    "extrapolated_beatstep": batch_ext_beatstep,
+                    "attention_mask_input_features": [None] * len(batch_input_feature),
+                    "attention_mask_beatsteps": [None] * len(batch_input_feature),
+                    "attention_mask_extrapolated_beatstep": [None] * len(batch_input_feature),
+                }
+            )
+        else:
+            output = BatchFeature(
+                {
+                    "input_features": batch_input_feature,
+                    "beatsteps": batch_beatsteps,
+                    "extrapolated_beatstep": batch_ext_beatstep,
+                }
+            )
+
+        if is_batched or return_attention_mask:
+            output = self.pad(output)
```

Comment:
I wonder whether we should pad the audio first then do all the spectrogram extraction? Rather than the other way round

Reply:
This is what we do in Whisper by the way: https://github.com/huggingface/transformers/blob/6b7d6f848bffb2ed68935858e71395c4070e3200/src/transformers/models/whisper/feature_extraction_whisper.py#L254-L256

Reply:
Even if we pad the audio files beforehand(to make them say of 300 secs each), the extract_rhythm will give different shapes of outputs even for each example! So we still need to pad them after that function.